chips/sam4l/src/usart.rs

```diff
@@ -276,6 +276,59 @@ const USART_BASE_ADDRS: [*mut UsartRegisters; 4] = [
     0x40030000 as *mut UsartRegisters,
 ];
 
+pub struct USARTRegManager<'a> {
+    registers: &'a UsartRegisters,
+    clock: pm::Clock,
+    rx_dma: Option<&'static dma::DMAChannel>,
+    tx_dma: Option<&'static dma::DMAChannel>,
+    is_panic: bool,
+}
+
+impl<'a> USARTRegManager<'a> {
+    fn real_new(usart: &USART, is_panic: bool) -> USARTRegManager {
+        unsafe {
+            if pm::is_clock_enabled(usart.clock) == false {
+                pm::enable_clock(usart.clock);
+            }
+        }
+        let regs: &UsartRegisters = unsafe { &*usart.registers };
+        USARTRegManager {
+            registers: regs,
+            clock: usart.clock,
+            rx_dma: usart.rx_dma.get(),
+            tx_dma: usart.tx_dma.get(),
+            is_panic: is_panic,
+        }
+    }
+
+    fn new(usart: &USART) -> USARTRegManager {
+        USARTRegManager::real_new(usart, false)
+    }
+
+    pub fn panic_new(usart: &USART) -> USARTRegManager {
+        USARTRegManager::real_new(usart, true)
+    }
+}
+
+impl<'a> Drop for USARTRegManager<'a> {
+    fn drop(&mut self) {
+        // Anything listening for RX or TX interrupts?
+        let ints_active = self.registers.imr.matches(
+            Interrupt::RXBUFF::SET + Interrupt::TXEMPTY::SET + Interrupt::TIMEOUT::SET
+                + Interrupt::PARE::SET + Interrupt::FRAME::SET + Interrupt::OVRE::SET
+                + Interrupt::TXRDY::SET + Interrupt::RXRDY::SET,
+        );
+
+        let rx_active = self.rx_dma.map_or(false, |rx_dma| rx_dma.is_enabled());
+        let tx_active = self.tx_dma.map_or(false, |tx_dma| tx_dma.is_enabled());
+        if !(rx_active || tx_active || ints_active || self.is_panic) {
```

Comment:
Why do we not want to disable the clock just because we are in a `panic!`? I agree that it is pointless to disable the clock, but it seems harmless to do so if the other checks pass. If there is an important reason, please add some comments to document it.